ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -340,3 +372,165 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
+  deriving (Generic, Show, Eq, Ord, Hashable, NFData)
+
+type BindingSigMap = Map.Map Id String
+
+newtype WhereBindingTypeSigsResult = WhereBindingTypeSigsResult ([WhereBindings], BindingSigMap)
+
+instance Show WhereBindingTypeSigsResult where
+  show _ = "<GetTypeResult.where>"
+
+instance NFData WhereBindingTypeSigsResult where
+  rnf = rwhnf
+
+type instance RuleResult GetWhereBindingTypeSigs = WhereBindingTypeSigsResult
+
+whereBindingRules :: Recorder (WithPriority Log) -> Rules ()
+whereBindingRules recorder = do
+  define (cmapWithPrio LogShake recorder) $ \GetWhereBindingTypeSigs nfp -> do
+    tmr <- use TypeCheck nfp
+    -- we need session here for tidying types
+    hsc <- use GhcSession nfp
+    result <- liftIO $ whereBindingType (tmrTypechecked <$> tmr) (hscEnv <$> hsc)
+    pure ([], result)
+
+whereBindingType :: Maybe TcGblEnv -> Maybe HscEnv -> IO (Maybe WhereBindingTypeSigsResult)
+whereBindingType (Just gblEnv) (Just hsc) = do
+  let wheres = findWhereQ (tcg_binds gblEnv)
+      localBindings = mapMaybe findBindingsQ wheres
+      bindToSig' = bindToSig hsc (tcg_rdr_env gblEnv)
+      findSigs (WhereBindings bindings _) = fmap findSig bindings
+        where findSig (WhereBinding bindingId _ _) = sequence (bindingId, bindToSig' bindingId)
+  (_, Map.fromList . fromMaybe [] -> sigMap) <-
+    initTcWithGbl hsc gblEnv ghostSpan $ sequence $ concatMap findSigs localBindings
+  pure $ Just (WhereBindingTypeSigsResult (localBindings, sigMap))
+whereBindingType _ _                   = pure Nothing
+
+-- | All where clauses from type checked source.
+findWhereQ :: GenericQ [HsLocalBinds GhcTc]
+findWhereQ = everything (<>) $ mkQ [] (pure . findWhere)
+  where
+    findWhere :: GRHSs GhcTc (LHsExpr GhcTc) -> HsLocalBinds GhcTc
+    findWhere = grhssLocalBinds
+
+-- | Find all bindings for **one** where clause.
+findBindingsQ :: GenericQ (Maybe WhereBindings)
+findBindingsQ = something (mkQ Nothing findBindings)
+  where
+    findBindings :: NHsValBindsLR GhcTc -> Maybe WhereBindings
+    findBindings (NValBinds binds sigs) =
+      Just $ WhereBindings
+        { bindings = concat $ mapMaybe (something (mkQ Nothing findBindingIds) . snd) binds
+        , existingSigNames = concatMap findSigIds sigs
+        }
+
+    findBindingIds :: LHsBindLR GhcTc GhcTc -> Maybe [WhereBinding]
+    findBindingIds bind = case unLoc bind of
+      FunBind{..} ->
+        let whereBinding = WhereBinding (unLoc fun_id) (getLoc fun_id)
+                           (col (getLoc fun_id) - col (getLoc bind))
+        in Just $ pure whereBinding
+      PatBind{..} -> Just $ (everything (<>) $ mkQ [] (fmap (uncurry wb) . maybeToList . findIdFromPat)) pat_lhs
+        where
+          wb id srcSpan = WhereBinding id srcSpan (col srcSpan - col (getLoc pat_lhs))
+      _           -> Nothing
+      where
+        col = srcSpanStartCol . realSrcSpan
+
+    -- | Example: Find `a` and `b` from @(a,b) = (1,True)@
```

Comment:
This doesn't seem quite right, it's only finding one thing, not multiple things?